hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ReadReplicas.java

```diff
@@ -0,0 +1,218 @@
+
+      for (Map.Entry<DatanodeDetails, OzoneInputStream>
+          replica : block.getValue().entrySet()) {
+        JsonObject replicaJson = new JsonObject();
+
+        replicaJson.addProperty("hostname",
+            replica.getKey().getHostName());
+        replicaJson.addProperty("uuid",
+            replica.getKey().getUuidString());
+
+        OzoneInputStream is = replica.getValue();
+        String fileName = keyName + "_block" + blockIndex + "_" +
+            replica.getKey().getHostName();
+        System.out.println("Writing : " + fileName);
+        File replicaFile
+            = new File("/opt/hadoop/" + directoryName + "/" + fileName);
+
+        try {
+          Files.copy(is, replicaFile.toPath(),
+              StandardCopyOption.REPLACE_EXISTING);
+        } catch (IOException e) {
+          Throwable cause = e.getCause();
+          replicaJson.addProperty("exception", e.getMessage());
+          if(cause instanceof OzoneChecksumException) {
+            BlockID blockID = block.getKey().getBlockID();
+            String datanodeUUID = replica.getKey().getUuidString();
+            is = getInputStreamWithoutChecksum(replicasWithoutChecksum,
+                datanodeUUID, blockID);
+            Files.copy(is, replicaFile.toPath(),
+                StandardCopyOption.REPLACE_EXISTING);
+          } else if(cause instanceof StatusRuntimeException) {
+            break;
+          }
+        } finally {
+          is.close();
+        }
+        replicasJson.add(replicaJson);
+      }
+      blockJson.add("replicas", replicasJson);
+      blocks.add(blockJson);
+    }
+  }
+
+  private OzoneInputStream getInputStreamWithoutChecksum(
+      Map<OmKeyLocationInfo, Map<DatanodeDetails, OzoneInputStream>>
+          replicasWithoutChecksum, String datanodeUUID, BlockID blockID) {
+    OzoneInputStream is = new OzoneInputStream();
+    for (Map.Entry<OmKeyLocationInfo, Map<DatanodeDetails, OzoneInputStream>>
+        block : replicasWithoutChecksum.entrySet()) {
+      if(block.getKey().getBlockID().equals(blockID)) {
+        for (Map.Entry<DatanodeDetails, OzoneInputStream>
+            replica : block.getValue().entrySet()) {
+          if(replica.getKey().getUuidString().equals(datanodeUUID)) {
+            is = replica.getValue();
+          }
+        }
+      }
+    }
+    return is;
+  }
+
+  @NotNull
+  private String createDirectory(String volumeName, String bucketName,
+                                 String keyName) {
+    String fileSuffix
+        = new SimpleDateFormat("yyyyMMddHHmmss").format(new Date());
+    String directoryName = volumeName + "_" + bucketName + "_" + keyName +
+        "_" + fileSuffix;
+    System.out.println("Creating directory : " + directoryName);
+    File dir = new File("/opt/hadoop/" + directoryName);
+    if (!dir.exists()){
+      if(dir.mkdir()) {
+        System.out.println("Successfully created!");
+      } else {
+        System.out.println("Something went wrong.");
```

Comment:
We might call this method at the very beginning of execution, and throw an exception here if the output directory can not be created, so that we fail early, and do not start gathering data that we can not save.

Reply:
Thanks, this is a good idea. I moved the execution earlier and changed the handling when the directory cannot be created.